contexts/PaymentContext/src/Domain/ChecksumGenerator.php

```diff
@@ -0,0 +1,43 @@
+<?php
+
+declare( strict_types = 1 );
+
+namespace WMDE\Fundraising\Frontend\PaymentContext\Domain;
+
+/**
+ * @licence GNU GPL v2+
+ */
+class ChecksumGenerator {
+
+	private $checksumCharacters;
+
+	/**
+	 * @param array $checksumCharacters Characters that can be used for the checksum
+	 */
+	public function __construct( array $checksumCharacters ) {
+		if ( count( $checksumCharacters ) < 2 ) {
+			throw new \InvalidArgumentException(
+				'Need at least two characters to create meaningful checksum'
+			);
+		}
+
+		$this->checksumCharacters = $checksumCharacters;
+	}
+
+	/**
+	 * @param string $string The string to create a checksum for
+	 *
+	 * @return string The checksum as a single character present in the constructors array argument
+	 */
+	public function createChecksum( string $string ): string {
+		$checksum = array_sum(
+			array_map(
+				'ord', // Returns the ASCII value of the character
+				str_split( md5( $string ) )
+			)
+		);
+
+		return $this->checksumCharacters[$checksum % count( $this->checksumCharacters )];
```

Comment:
This is brilliant because it makes the character selection independent from the bit-length of the checksum.

Reply:
Do I get an achievement card?

Reply:
I don't have pretty cards, but here you go ...

![img_4563](https://user-images.githubusercontent.com/223326/30689848-8adc05c6-9ec2-11e7-9c93-2511e0884802.JPG)